javalib/src/main/scala/java/net/URI.scala

```diff
@@ -601,39 +614,91 @@ object URI {
 
   // Quote helpers
 
-  private val quoteChar: js.Function1[String, String] = { (str: String) =>
-    require(str.length == 1)
+  private def decodeComponent(str: String): String = {
+    val inBuf = CharBuffer.wrap(str)
+    val outBuf = CharBuffer.allocate(inBuf.capacity)
+    val byteBuf = ByteBuffer.allocate(64)
+    val cBuf = new Array[Char](2)
+    val decoder = StandardCharsets.UTF_8.newDecoder
+      .onMalformedInput(CodingErrorAction.REPLACE)
+      .onUnmappableCharacter(CodingErrorAction.REPLACE)
+
+    def decode(endOfInput: Boolean) = {
+      byteBuf.flip()
+      decoder.decode(byteBuf, outBuf, endOfInput)
+      if (endOfInput) {
+        decoder.reset()
+        byteBuf.clear()
+      } else {
+        byteBuf.compact()
+      }
+    }
 
-    val c = str.head.toInt
+    while (inBuf.hasRemaining) {
+      inBuf.get match {
+        case '%' =>
+          if (!byteBuf.hasRemaining)
+            decode(false)
 
-    if (c > 127)
-      throw new URISyntaxException(null, "Only ASCII allowed in URIs")
-    else
-      f"%%$c%02x"
+          inBuf.get(cBuf) // get two chars
+          val hexStr = new String(cBuf)
+          val v = Integer.parseInt(hexStr, 16)
+          byteBuf.put(v.toByte)
+
+        case c =>
+          decode(true)
+          outBuf.put(c)
+      }
+    }
+
+    decode(true)
+
+    outBuf.flip()
+    outBuf.toString
+  }
+
+  private val quoteStr: js.Function1[String, String] = { (str: String) =>
+    val buf = StandardCharsets.UTF_8.encode(str)
+
+    var res = ""
+    while (buf.hasRemaining) {
+      val c = buf.get & 0xff
+      res += f"%%$c%02X"
```

Comment:
I'd use this occasion to get rid of this unnecessary inefficient incantation:
```scala
res += (if (c <= 0xf) "%0" else "%") + Integer.toHexString(c).toUpperCase
```
since now, we're more likely to encode quite a few characters per URI construction.